client/state/site-settings/actions.js

```diff
@@ -0,0 +1,109 @@
+/**
+ * Internal dependencies
+ */
+import wpcom from 'lib/wp';
+import {
+	SITE_SETTINGS_RECEIVE,
+	SITE_SETTINGS_REQUEST,
+	SITE_SETTINGS_REQUEST_FAILURE,
+	SITE_SETTINGS_REQUEST_SUCCESS,
+	SITE_SETTINGS_SAVE,
+	SITE_SETTINGS_SAVE_FAILURE,
+	SITE_SETTINGS_SAVE_SUCCESS,
+	SITE_SETTINGS_UPDATE
+} from 'state/action-types';
+
+/**
+ * Returns an action object to be used in signalling that site settings have been received.
+ *
+ * @param  {Number} siteId Site ID
+ * @param  {Object} settings The site settings object
+ * @return {Object}        Action object
+ */
+export function receiveSiteSettings( siteId, settings ) {
+	return {
+		type: SITE_SETTINGS_RECEIVE,
+		siteId,
+		settings
+	};
+}
+
+/**
+ * Returns an action object to be used in signalling that some site settings have been update.
+ *
+ * @param  {Number} siteId Site ID
+ * @param  {Object} settings The updated site settings
+ * @return {Object}        Action object
+ */
+export function updateSiteSettings( siteId, settings ) {
+	return {
+		type: SITE_SETTINGS_UPDATE,
+		siteId,
+		settings
+	};
+}
+
+/**
+ * Returns an action thunk which, when invoked, triggers a network request to
+ * retrieve site settings
+ *
+ * @param  {Number} siteId Site ID
+ * @return {Function}      Action thunk
+ */
+export function requestSiteSettings( siteId ) {
+	return ( dispatch ) => {
+		dispatch( {
+			type: SITE_SETTINGS_REQUEST,
+			siteId
+		} );
+
+		return wpcom.undocumented().settings( siteId )
+			.then( ( { name, description, settings } ) => {
+				const savedSettings = {
+					...settings,
+					blogname: name,
+					blogdescription: description
+				};
+
+				dispatch( receiveSiteSettings( siteId, savedSettings ) );
+				dispatch( {
+					type: SITE_SETTINGS_REQUEST_SUCCESS,
+					siteId
+				} );
+			} )
+			.catch( error => {
+				dispatch( {
+					type: SITE_SETTINGS_REQUEST_FAILURE,
+					siteId,
+					error
+				} );
+			} );
+	};
+}
+
+export function saveSiteSettings( siteId, settings ) {
+	return ( dispatch ) => {
+		dispatch( {
```

Comment:
It's an interesting example. Do you think there's harm in optimistic updating, if it will eventually fix itself in the end after the response is received? There seem to be fewer exceptions anyways, and perhaps we can recreate some in the client (`gmt_offset`, for example, is [generated in the response](https://github.com/Automattic/jetpack/blob/edfc98cabd39f45ac6f5b7d288e97b24bc481d5f/json-endpoints/class.wpcom-json-api-site-settings-endpoint.php#L409-L418) via a simple regular expression match).